forge-core/src/main/java/forge/util/collect/ConcurrentMultiMap.java

```diff
@@ -0,0 +1,135 @@
+package forge.util.collect;
+
+import com.google.common.collect.ConcurrentHashMultiset;
+import com.google.common.collect.Multiset;
+
+import java.util.Collection;
+import java.util.Map;
+import java.util.Queue;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentLinkedQueue;
+
+public class ConcurrentMultiMap<K, V> {
```

Comment:
Why exactly using your on class instead of using the collection4 package?
It wouldn't be much of a problem, I was probably going to use something of that package anyway

And should this implement https://guava.dev/releases/23.0/api/docs/com/google/common/collect/Multimap.html ?

Reply:
I tried implementing the guava multimap but you need to fulfill all those overrides (seems overkill for the manapool sake version), though the earlier version from collection4 package works too.

Reply:
To me the question is why does ManaPool need to be concurrent now?

If you're wanting to check multiple SA for p(l)ayability in some non-deterministic way won't some checks possibly fail if the pool is already being drained from another Future/Thread at the same time?!

Reply:
I thought we only get a single spellability to play, the only difference is they're computed concurrently but ai will play at a time, by the time the ai cast or activate the top sa, wouldn't the next sa checked again if feasible to play? If you have better idea to access the manapool without concurrent modification then please enlighten.

Reply:
- checking SA includes mana test
- if more than one is available and the pool isn't empty then how are you keeping it synchronized when AI checks A + B at roughly the same time? **maybe you didn't consider it but the pool is actively being modified/drained even when only testing**

the whole AI logic isn't set up to jump around between threads even if the data structures might not crash from it now lol

Reply:
Would it be cleaner + efficient enough to implement threading through something more like the GameSimulator AI - copy the whole game state to separate threads and evaluate them without any risk of tampering with shared resources?